### PR TITLE
Added 'Laptop Price Predictor' Project

### LINKED ISSUE #1029: Laptop Price Predictor
**Is your feature request related to a problem? Please describe.**
The task of buying a new laptop can be very hectic. Comparing the companies, prices, model features, etc. is very time taking.

**Describe the solution you'd like**
My project can ease this process by using ML. A nice catchy GUI will make the process even more fun!

**Additional context**
My project folder will be complete with the dataset, the notebook, model's pkl file, the GUI's notebook and a descriptive README.

The final GUI will look something like this:

image

### PR DESCRIPTION
**Pull request by Manasi2001**
# Description

The folder structure goes like this:

Laptop Price Predictor

▫Dataset
- laptoppricing.csv

▫GUI
- app.py

▫Model
- Laptop_Price_Prediction.py
- LaptopPricePrediction.pkl

▫README

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)

## Project Metadata

`` If there is no-file/nothing to fill the below fields with, then type: none ``

`` Example: `` If no requirements.txt needed/present, then type none in ``Requirments``.

Category:
- [ ] Calculators
- [x] AI/ML
- [ ] Scrappers
- [ ] Social_Media
- [ ] PDF
- [ ] Image_Processing
- [ ] Video_Processing
- [ ] Games
- [ ] Networking
- [ ] OS_Utilities
- [ ] Automation
- [ ] Cryptography
- [ ] Computer_Vision
- [ ] Fun
- [ ] Others

Title: Laptop Price Predictor

Folder: Model

Requirements: None

Script: Laptop_Price_Prediction.py

Arguments: None

Contributor: Manasi2001

Description: This 'Laptop Price Predictor' provided with a UI, can very accurately predict the prices of laptops on the basis of companies, model features, etc.

image

### CODE REVIEW

File: Laptop Price Predictor/GUI/app.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import joblib
+model = joblib.load(r'C:\Users\DELL\Desktop\Kaggle+HE\Github GSSoC21\Amazing Python Scripts\Laptop Price Predictor\Model\LaptopPricePrediction.pkl')
+
```
Comment (santushtisharma10) [inline Laptop Price Predictor/GUI/app.py:3]:
```suggestion
model = joblib.load(r'./Laptop Price Predictor/Model/LaptopPricePrediction.pkl')

```

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from sklearn import preprocessing
+from statsmodels.stats.outliers_influence import variance_inflation_factor
+from sklearn.model_selection import train_test_split
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.metrics import r2_score
+
+# Getting our Data
+
+df = pd.read_csv(r'C:\Users\DELL\Desktop\Kaggle+HE\Github GSSoC21\Amazing Python Scripts\Laptop Price Predictor\Dataset\laptoppricing.csv', encoding = 'unicode_escape')
+
```
Comment (santushtisharma10) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:15]:
```suggestion

df = pd.read_csv(r'./Laptop Price Predictor/Dataset/laptoppricing.csv', encoding = 'unicode_escape')

```

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+y_pred = regr.predict(X_test)
+
+print('\nAccuracy:',r2_score(y_test, y_pred))
+
+# Predictions are 98.79% accurate.
+
+# Getting the pkl file
+
+import pickle
+pickle.dump(regr, open('LaptopPricePrediction.pkl', 'wb'))
```
Comment (santushtisharma10) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:62]:
```suggestion
# Getting the pkl file

import pickle
pickle.dump(regr, open('./Laptop Price Predictor/Model/LaptopPricePrediction.pkl', 'wb'))
```

**Review by santushtisharma10 — CHANGES_REQUESTED** [review]

**Comment by santushtisharma10** [comment]
Also include the case if user enters any other brand than ASUS, DELL, HP, Lenovo .

**Comment by santushtisharma10** [comment]
@Manasi2001 instead of closing the application you should display a statement in the result like price can't be predicted due to lack of data of given model/manufacturer.

**Review by antrikshmisri — APPROVED** [review]

**Comment by Manasi2001** [comment]
@santushtisharma10 please check now.

File: Laptop Price Predictor/README.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+A Random Forest is an ensemble technique capable of performing both regression and classification tasks with the use of multiple decision trees and a technique called Bootstrap and Aggregation, commonly known as bagging. The basic idea behind this is to combine multiple decision trees in determining the final output rather than relying on individual decision trees.
+Random Forest has multiple decision trees as base learning models. We randomly perform row sampling and feature sampling from the dataset forming sample datasets for every model. This part is called Bootstrap.
+
+## Package Used for GUI
+
+tkinter
+
+## Author
+
+Manasi Chhibber
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/README.md:20]:
```suggestion
[Manasi Chhibber](https://github.com/Manasi2001)
```

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+# Getting our Data
+
+df = pd.read_csv(r'./Laptop Price Predictor/Dataset/laptoppricing.csv', encoding = 'unicode_escape')
+
+# Data Preprocessing
+
+# checking for null values
+print('\n',df.isnull().any())
+
+print('\n',df.columns)
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:21]: you can remove this

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+label_encoder = preprocessing.LabelEncoder()
+df['ï»¿Manufacturer'] = df['ï»¿Manufacturer'].astype('|S')
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:24]: what is this character?: `ï»¿`Manufacturer 

if the column is not clean, then add the step to clean it

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+# checking vif
+variables = df[['ï»¿Manufacturer', 'IntelCore(i-)', 'IntelCoreGen',
+       'processing speed(GHz)', 'Ram(gb)', 'HDD(gb)', 'SSD(gb)',
+       'Graphics(gb)', 'ScreenSize(inch)']]
+vif = pd.DataFrame()
+vif['VIF'] = [variance_inflation_factor(variables.values, i) for i in range(variables.shape[1])]
+vif['Features'] = variables.columns
+
+print('\n',vif)
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:35]:
```suggestion
print('Variance inflation factor is: ',vif)
```

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+# Model
+
+regr = RandomForestRegressor(random_state=0)
+regr.fit(X, y)
+
+# Making Predictions and Checking Accuracy
+
+y_pred = regr.predict(X_test)
+
+print('\nAccuracy:',r2_score(y_test, y_pred))
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:55]:
```suggestion
print('\nAccuracy (r2_score) :  ',r2_score(y_test, y_pred))
```

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+# Predictions are 98.79% accurate.
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:57]: remove this

File: Laptop Price Predictor/GUI/app.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
+import joblib
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/GUI/app.py:1]: shift this file to `Laptop Price Predictor` folder

File: Laptop Price Predictor/GUI/app.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    global mf # our 1st input variable
+    mf = (entry1.get()) 
+    if mf == 'ASUS':
+        m = 0
+    elif mf == 'Dell':
+        m = 1
+    elif mf == 'HP':
+        m = 2
+    elif mf == 'Lenovo':
+        m = 3
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/GUI/app.py:54]: can we have a dropdown for the manufacturer? It will eliminate the user error for the wrong manufacturer

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

**Review by kaustubhgupta — COMMENTED** [review]
add the requirements.txt file

**Review by santushtisharma10 — APPROVED** [review]

File: Laptop Price Predictor/app.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
+import joblib
+model = joblib.load('..\Laptop Price Predictor\Model\LaptopPricePrediction.pkl')
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/app.py:2]:
```suggestion
model = joblib.load('.\Laptop Price Predictor\Model\LaptopPricePrediction.pkl')
```

File: Laptop Price Predictor/Model/Laptop_Price_Prediction.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.metrics import r2_score
+
+# Getting our Data
+
+df = pd.read_csv(r'./Laptop Price Predictor/Dataset/laptoppricing.csv', encoding = 'unicode_escape')
+
+# Data Preprocessing
+
+df.rename(columns={'ï»¿Manufacturer':'Manufacturer'}, inplace=True)
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/Model/Laptop_Price_Prediction.py:18]: on writing this simple code, I am not getting any special characters for manufacturer:

```python
import pandas as pd

df = pd.read_csv(r"Laptop Price Predictor\Dataset\laptoppricing.csv")
print(df.columns)
```
Output:
```bash
Index(['Manufacturer', 'IntelCore(i-)', 'IntelCoreGen',
       'processing speed(GHz)', 'Ram(gb)', 'HDD(gb)', 'SSD(gb)',
       'Graphics(gb)', 'ScreenSize(inch)', 'Price'],
      dtype='object')
```
Just read the dataset without any encoding

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
Drop down menu for manufacturer is still missing. Also:

File: Laptop Price Predictor/app.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    r = float(entry2.get()) 
+
+    global s # our 3rd input variable
+    s = float(entry3.get())
+
+    global g # our 4th input variable
+    g = float(entry4.get())
+    
+    Prediction_result  = ('Predicted Laptop Price: ', model.predict([[m, r, s, g]]))
+    label_Prediction = tk.Label(root, text= Prediction_result, bg='orange')
```
Comment (kaustubhgupta) [inline Laptop Price Predictor/app.py:70]: This is not coming out attractive:

image

Curly braces with square brackets, just display the price value with appropriate currency symbol

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

**Comment by kaustubhgupta** [comment]
@Manasi2001 Updates on this?

**Comment by Manasi2001** [comment]
> @Manasi2001 Updates on this?

I'm not willing to work on this anymore @kaustubhgupta.

**Pull request closed without merging**